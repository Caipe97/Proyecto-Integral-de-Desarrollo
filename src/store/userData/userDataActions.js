import {
  LOGOUT,
  REFRESH_PAGE,

  LOGIN_OR_REGISTER_PENDING,
  LOGIN_OR_REGISTER_SUCCESS,
  LOGIN_OR_REGISTER_FAILED,

  RESET_PASSWORD_PENDING,
  RESET_PASSWORD_SUCCESS,
  RESET_PASSWORD_FAILED
 } from './userDataConstants'

export const logout = () => ({ 
  type: LOGOUT
})

export const refreshPage = () => ({ 
  type: REFRESH_PAGE
})

export const login = (email, password) => (dispatch) => {
  dispatch({ type: LOGIN_OR_REGISTER_PENDING })
  return(
    fetch(`https://jma-test-app.herokuapp.com/api/users/login`, {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({
        email: email,
        password: password
      })
    })
  )
  .then(response => response.json())
  .then(userData => dispatch({ type: LOGIN_OR_REGISTER_SUCCESS, payload: userData }))
  .catch(error => dispatch({ type: LOGIN_OR_REGISTER_FAILED, payload: error }))
}

export const register = (name, surname, email, password, gender, birthday, weight, height) => (dispatch) => {
  dispatch({ type: LOGIN_OR_REGISTER_PENDING });
  return(
    fetch(`https://jma-test-app.herokuapp.com/api/users/register`, {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({
        name: name,
        surname: surname,
        email: email,
        password: password,
        gender: gender,
        birthday: birthday,
        weight: weight,
        height: height
      })
    })
  )
  .then(response => response.json())
  .then(userData => dispatch({ type: LOGIN_OR_REGISTER_SUCCESS, payload: userData }))
  .catch(error => dispatch({ type: LOGIN_OR_REGISTER_FAILED, payload: error }))
}

export const resetPassword = (userId, password) => (dispatch) => {
  dispatch({ type: RESET_PASSWORD_PENDING });
  return(
    fetch(`https://jma-test-app.herokuapp.com/api/users?userId=${userId}`, {
      method: 'PUT',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({
        password: password
      })
    })
  )
  .then(response => response.json())
  .then(userData => dispatch({ type: RESET_PASSWORD_SUCCESS, payload: userData }))
  .catch(error => dispatch({ type: RESET_PASSWORD_FAILED, payload: error }))
}