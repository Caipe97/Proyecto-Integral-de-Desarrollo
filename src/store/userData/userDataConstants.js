export const LOGOUT = 'LOGOUT';

export const LOGIN_OR_REGISTER_PENDING = 'LOGIN_OR_REGISTER_PENDING';
export const LOGIN_OR_REGISTER_SUCCESS = 'LOGIN_OR_REGISTER_SUCCESS';
export const LOGIN_OR_REGISTER_FAILED = 'LOGIN_OR_REGISTER_FAILED';

export const RESET_PASSWORD_PENDING = 'RESET_PASSWORD_PENDING';
export const RESET_PASSWORD_SUCCESS = 'RESET_PASSWORD_SUCCESS';
export const RESET_PASSWORD_FAILED = 'RESET_PASSWORD_FAILED';