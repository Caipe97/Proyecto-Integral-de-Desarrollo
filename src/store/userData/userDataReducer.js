import {
  LOGOUT,
  REFRESH_PAGE,

  LOGIN_OR_REGISTER_PENDING,
  LOGIN_OR_REGISTER_SUCCESS,
  LOGIN_OR_REGISTER_FAILED,

  RESET_PASSWORD_PENDING,
  RESET_PASSWORD_SUCCESS,
  RESET_PASSWORD_FAILED
 } from './userDataConstants';

const initialStateUserData = {
  userId: 0,
  name: '',
  surname: '',
  email: '',
  password: '',
  gender: '',
  birthday: '',
  weight: '',
  height: '',
  isPending: false
}

export const userDataReducer = (state=initialStateUserData, action={}) => {
  switch (action.type) {
    case LOGOUT:
      localStorage.clear();
      return {
        userId: 0,
        name: '',
        surname: '',
        email: '',
        password: '',
        gender: '',
        birthday: '',
        weight: '',
        height: '',
        isPending: false
      }
    case REFRESH_PAGE:
      const userData = JSON.parse(localStorage.getItem('userData'))
      return {
        userId: userData.userId,
        name: userData.name,
        surname: userData.surname,
        email: userData.email,
        password: userData.password,
        gender: userData.gender,
        birthday: userData.birthday,
        weight: userData.weight,
        height: userData.height,
        isPending: false
      }
    case LOGIN_OR_REGISTER_PENDING:
      return {
        ...state,
        isPending: true
      }
    case LOGIN_OR_REGISTER_SUCCESS:
      localStorage.setItem('userData', JSON.stringify(action.payload));
      return {
        ...state,
        userId: action.payload.userId,
        name: action.payload.name,
        surname: action.payload.surname,
        email: action.payload.email,
        password: action.payload.password,
        gender: action.payload.gender,
        birthday: action.payload.birthday,
        weight: action.payload.weight,
        height: action.payload.height,
        isPending: false
      }
    case LOGIN_OR_REGISTER_FAILED:
      return {
        ...state,
        error: action.payload,
        isPending: false
      }
    case RESET_PASSWORD_PENDING:
      return {
        ...state,
        isPending: true
      }
    case RESET_PASSWORD_SUCCESS:
      return {
        ...state,
        isPending: false
      }
    case RESET_PASSWORD_FAILED:
      return {
        ...state,
        isPending: false,
        error: action.payload
      }
    default:
      return state
  }
}