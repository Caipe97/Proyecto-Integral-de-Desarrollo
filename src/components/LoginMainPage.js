import React, { Component } from 'react';
import { Link } from "react-router-dom";
import '../pages/Login/Login.css';
import logo1 from '../images/avatar.png';
import logo2 from '../images/2pages.jpg';
import Header from './Header';
import Footer from './Footer';
import TextField from '@material-ui/core/TextField';
import { LOGIN_OR_REGISTER_SUCCESS } from '../store/userData/userDataConstants';

/* eslint-disable */
class LoginMainPage extends Component{
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: ''
    };
  }

  handleChange = event => {
    this.setState({
      ...this.state,
      [event.target.name]: event.target.value
    })
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    if(this.validateAll()) {
      const data = await this.props.onLogin(this.state.email, this.state.password);
      this.setState({email: '', password: ''});
      if(data.type === LOGIN_OR_REGISTER_SUCCESS){
        this.props.history.push("/profile");
      } else{
        console.log("Email y/o contraseÃ±a incorrectos");
      }
    } else {
      console.log('No completaste los campos correctamente');
    }
  };

  validateAll = () => {
    if(this.validateEmail(this.state.email) && this.validatePassword(this.state.password)) {
      return true;
    }
    return false;
  }

  validateEmail = (email) => {
    const expression = /(?!.*\.{2})^([a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+(\.[a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+)*|"((([\t]*\r\n)?[\t]+)?([\x01-\x08\x0b\x0c\x0e-\x1f\x7f\x21\x23-\x5b\x5d-\x7e\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|\\[\x01-\x09\x0b\x0c\x0d-\x7f\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))*(([\t]*\r\n)?[\t]+)?")@(([a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.)+([a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.?$/i;
    return expression.test(String(email).toLowerCase());
  }

  validatePassword = (password) => {
    const expression = /(.|\s)*\S(.|\s)*/ 
    return expression.test(String(password).toLowerCase())
  }

  render() {
    return (
      <div className="contenedorL">
        <Header {...this.props}/>

        <div className="contenidoL">
          <div className="col1">
          <img src={logo2} alt="imagen tracking" style={{ width:240,height:240 }} />
          <div>
          <h1 style={{left:'30px'}}>Nutriva</h1>
          </div>
          
          </div>
          
          <div className="col2" style ={{justifyContent:'center',alignContent:'center',alignItems:'center',textAlign:'center'}}>
          <p >
          <br></br>
          <br></br>
          <br></br>
          <br></br>
          <br></br>
          <br></br>
          <br></br>
          
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          
          </p>
          </div>
          
        </div>
        <div className="sidebar">
          <div style={{ display: 'flex', justifyContent: 'center', alignItems: 'center', flexDirection: 'column' }}>
            <div className="box" style={{}}>
              <img src={logo1} alt="imagen tracking" style={{ width: 100 }} />
            </div>
            <div style={{ marginTop: 45 }}>
              <h1 className='f1'>Login</h1>
            </div>

            <form style={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100%', flexDirection: 'column' }}>
              <TextField label="Email" name='email' type='email' value={this.state.email} onChange={this.handleChange} required/>
              <TextField label="Password" name='password' type='password' value={this.state.password} onChange={this.handleChange} required/>
              <button onClick={this.handleSubmit} className='button'>Iniciar Sesion</button>
              <p style={{marginBlock: '0em', marginTop: '4%'}}>Todavia no tenes una cuenta?
                <Link to="/register" style={{color: 'black'}}>Registrate</Link>
              </p>
              <p style={{fontSize:'10px',marginTop:'4%'}}>
                Olvidaste la contrasena?
              </p>
            </form>
          </div>
        </div>
      
        <Footer/>
      </div>

    );
  }
}

export default LoginMainPage
