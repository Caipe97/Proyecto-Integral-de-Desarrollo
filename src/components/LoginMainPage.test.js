import React from 'react';
import { shallow } from 'enzyme';
import LoginMainPage from './LoginMainPage';
import { LOGIN_OR_REGISTER_SUCCESS, LOGIN_OR_REGISTER_FAILED } from '../store/userData/userDataConstants';

describe('LoginMainPage tests', () => {
    const historyMock = { push: jest.fn() };
    const preventDefault = { preventDefault: jest.fn() }
    let wrapper;
    const onLoginReturnedDataSuccess = { 
        type: LOGIN_OR_REGISTER_SUCCESS,
        payload: {
            userId: 1,
            name: 'julian',
            surname: 'livrone',
            email: 'julianlivrone@gmail.com',
            password: 'asd',
            gender: 'masculino',
            birthday: '2010-01-21',
            weight: 90,
            height: 170
        }
    };

    beforeEach(() => {
        let mockProps = {
            userDataReducer: {
                userId: 0,
                name: '',
                surname: '',
                email: '',
                password: '',
                gender: '',
                birthday: '',
                weight: '',
                height: '',
                isPending: false
            },
            onLogin: jest.fn().mockReturnValueOnce(onLoginReturnedDataSuccess),
            onLogout: jest.fn(),
            preventDefault:  jest.fn()
          };
        wrapper = shallow(<LoginMainPage {...mockProps} history={historyMock}/>);
    })

    
    it('renders LoginMainPage without crashing', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    it('handleChange should update the state correctly when writing on the inputs', () => {
        wrapper.find('[name="email"]').at(0).simulate('change', { target: { name: 'email', value: 'julianlivrone@gmail.com' } });
        wrapper.find('[name="password"]').at(0).simulate('change', { target: { name: 'password', value: 'asd' } });
        expect(wrapper.instance().state).toEqual({email: 'julianlivrone@gmail.com', password: 'asd'});
    })

    it('handleSubmit should update the state to its initial state (empty)', async () => {
        wrapper.find('[name="email"]').at(0).simulate('change', { target: { name: 'email', value: 'julianlivrone@gmail.com' } });
        wrapper.find('[name="password"]').at(0).simulate('change', { target: { name: 'password', value: 'asd' } });
        await wrapper.find('[className="button"]').at(0).simulate('click', preventDefault);
        expect(wrapper.instance().state).toEqual({email: '', password: ''});
    })

    it('handleSubmit should push to profile when fetching the userData when email and password are correct', async () => {
        wrapper.find('[name="email"]').at(0).simulate('change', { target: { name: 'email', value: 'julianlivrone@gmail.com' } });
        wrapper.find('[name="password"]').at(0).simulate('change', { target: { name: 'password', value: 'asd' } });
        await wrapper.find('[className="button"]').at(0).simulate('click', preventDefault);
        expect(wrapper.instance().props.onLogin).toHaveBeenCalledTimes(1);
        expect(wrapper.instance().props.history.push).toHaveBeenCalledTimes(1);
        expect(historyMock.push.mock.calls[0]).toEqual(['/profile']);
    })

    it('handleSubmit should not push to profile because email is not valuserId 1', async () => {
        wrapper.find('[name="email"]').at(0).simulate('change', { target: { name: 'email', value: 'julianlivrone@' } });
        wrapper.find('[name="password"]').at(0).simulate('change', { target: { name: 'password', value: 'asd' } });
        await wrapper.find('[className="button"]').at(0).simulate('click', preventDefault);
        expect(wrapper.instance().props.onLogin).toHaveBeenCalledTimes(0);
        expect(wrapper.instance().props.history.push).toHaveBeenCalledTimes(0);
    })

    it('handleSubmit should not push to profile because email is not valuserId 2', async () => {
        wrapper.find('[name="email"]').at(0).simulate('change', { target: { name: 'email', value: 'julianlivrone@dsa' } });
        wrapper.find('[name="password"]').at(0).simulate('change', { target: { name: 'password', value: 'asd' } });
        await wrapper.find('[className="button"]').at(0).simulate('click', preventDefault);
        expect(wrapper.instance().props.onLogin).toHaveBeenCalledTimes(0);
        expect(wrapper.instance().props.history.push).toHaveBeenCalledTimes(0);
    })

    it('handleSubmit should not push to profile because password is not valuserId 1', async () => {
        wrapper.find('[name="email"]').at(0).simulate('change', { target: { name: 'email', value: 'julianlivrone@gmail.com' } });
        wrapper.find('[name="password"]').at(0).simulate('change', { target: { name: 'password', value: '' } });
        await wrapper.find('[className="button"]').at(0).simulate('click', preventDefault);
        expect(wrapper.instance().props.onLogin).toHaveBeenCalledTimes(0);
        expect(wrapper.instance().props.history.push).toHaveBeenCalledTimes(0);
    })

    it('handleSubmit should not push to profile when cannot fetch userData because email or password are incorrect', async () => {
        const onLoginReturnedDataFailed = { 
            type: LOGIN_OR_REGISTER_FAILED,
            payload: {
                message: 'No user found with given credentials'
            }
        };
        let mockProps = {
            userDataReducer: {
                userId: 0,
                name: '',
                surname: '',
                email: '',
                password: '',
                gender: '',
                birthday: '',
                weight: '',
                height: '',
                isPending: false
            },
            // onLogin: jest.fn().mockRejectedValueOnce("ERROR: could not fetch data"),
            onLogin: jest.fn().mockResolvedValueOnce(onLoginReturnedDataFailed),
            onLogout: jest.fn()
          };
        wrapper = shallow(<LoginMainPage {...mockProps} history={historyMock}/>);
        wrapper.find('[name="email"]').at(0).simulate('change', { target: { name: 'email', value: 'julianlivrone@gmail.com' } });
        wrapper.find('[name="password"]').at(0).simulate('change', { target: { name: 'password', value: 'as' } });
        await wrapper.find('[className="button"]').at(0).simulate('click', preventDefault);
        expect(wrapper.instance().props.onLogin).toHaveBeenCalledTimes(1);
        expect(wrapper.instance().props.history.push).toHaveBeenCalledTimes(0);
    })
})