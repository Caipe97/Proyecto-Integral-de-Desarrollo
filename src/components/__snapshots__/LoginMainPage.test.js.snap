// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginMainPage tests renders LoginMainPage without crashing 1`] = `
<div
  className="contenedorL"
>
  <Header
    history={
      Object {
        "push": [MockFunction],
      }
    }
    onLogin={[MockFunction]}
    onLogout={[MockFunction]}
    preventDefault={[MockFunction]}
    userDataReducer={
      Object {
        "birthday": "",
        "email": "",
        "gender": "",
        "height": "",
        "id": 0,
        "isPending": false,
        "name": "",
        "password": "",
        "surname": "",
        "weight": "",
      }
    }
  />
  <div
    className="contenidoL"
  >
    <div
      className="col1"
    >
      <img
        alt="imagen tracking"
        src="2pages.jpg"
        style={
          Object {
            "height": 240,
            "width": 240,
          }
        }
      />
      <div>
        <h1
          style={
            Object {
              "left": "30px",
            }
          }
        >
          Nutriva
        </h1>
      </div>
    </div>
    <div
      className="col2"
      style={
        Object {
          "alignContent": "center",
          "alignItems": "center",
          "justifyContent": "center",
          "textAlign": "center",
        }
      }
    >
      <p>
        <br />
        <br />
        <br />
        <br />
        <br />
        <br />
        <br />
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </p>
    </div>
  </div>
  <div
    className="sidebar"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <div
        className="box"
        style={Object {}}
      >
        <img
          alt="imagen tracking"
          src="avatar.png"
          style={
            Object {
              "width": 100,
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "marginTop": 45,
          }
        }
      >
        <h1
          className="f1"
        >
          Login
        </h1>
      </div>
      <form
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "center",
          }
        }
      >
        <WithStyles(ForwardRef(TextField))
          label="Email"
          name="email"
          onChange={[Function]}
          required={true}
          type="email"
          value=""
        />
        <WithStyles(ForwardRef(TextField))
          label="Password"
          name="password"
          onChange={[Function]}
          required={true}
          type="password"
          value=""
        />
        <button
          className="button"
          onClick={[Function]}
        >
          Iniciar Sesion
        </button>
        <p
          style={
            Object {
              "marginBlock": "0em",
              "marginTop": "4%",
            }
          }
        >
          Todavia no tenes una cuenta?
          <Link
            style={
              Object {
                "color": "black",
              }
            }
            to="/register"
          >
            Registrate
          </Link>
        </p>
        <p
          style={
            Object {
              "fontSize": "10px",
              "marginTop": "4%",
            }
          }
        >
          Olvidaste la contrasena?
        </p>
      </form>
    </div>
  </div>
  <Footer />
</div>
`;
